overlord/snapstate/snapstate.go

```diff
@@ -4146,6 +4148,120 @@ func GadgetConnections(st *state.State, deviceCtx DeviceContext) ([]gadget.Conne
 	return gadgetInfo.Connections, nil
 }
 
+// downloadsToKeep returns a map of download file names that need to be kept
+// for all current snaps in the system state.
+//
+// A downloaded file is only kept if any of the following are true:
+//    1. The revision is in SnapState.Sequence
+//    2. The revision is in saved in refresh-candidates
+//    3. The revision is pointed to by a download task in an ongoing change
+//
+// It is the caller's responsibility to lock state before calling this function.
+func downloadsToKeep(st *state.State) (map[string]bool, error) {
+	snapStates, err := All(st)
+	if err != nil {
+		return nil, err
+	}
+
+	// initialize refresh hints
+	var refreshHints map[string]*refreshCandidate
+	if err := st.Get("refresh-candidates", &refreshHints); err != nil {
+		if !errors.Is(err, &state.NoStateError{}) {
+			return nil, fmt.Errorf("cannot get refresh-candidates: %v", err)
+		}
+	}
+
+	downloadsToKeep := map[string]bool{}
+	keep := func(name string, rev snap.Revision) {
+		downloadsToKeep[fmt.Sprintf("%s_%s.snap", name, rev)] = true
+	}
+
+	// keep revisions in snap's sequence
+	for snapName, snapst := range snapStates {
+		for _, seq := range snapst.Sequence {
+			keep(snapName, seq.Revision)
+		}
+	}
+
+	// keep revisions in refresh hints
+	for snapName, hint := range refreshHints {
+		keep(snapName, hint.Revision())
+	}
+
+	// keep revisions pointed to by a download task in an ongoing change
+	for _, chg := range st.Changes() {
+		if chg.IsReady() {
+			continue
+		}
+		for _, t := range chg.Tasks() {
+			if t.Kind() != "download-snap" {
+				continue
+			}
+			snapsup, err := TaskSnapSetup(t)
+			if err != nil {
+				return nil, err
+			}
+			keep(snapsup.InstanceName(), snapsup.Revision())
+		}
+	}
+
+	return downloadsToKeep, nil
+}
+
+// cleanDownloads removes downloads that are no longer needed for all snaps.
+//
+// It is the caller's responsibility to lock state before calling this function.
+var cleanDownloads = func(st *state.State) error {
+	keep, err := downloadsToKeep(st)
+	if err != nil {
+		return err
+	}
+
+	matches, err := filepath.Glob(filepath.Join(dirs.SnapBlobDir, "*.snap"))
+	if err != nil {
+		return err
+	}
+	for _, file := range matches {
+		if keep[filepath.Base(file)] {
+			continue
+		}
+		if rmErr := os.Remove(file); rmErr != nil {
+			// continue deletion, report error in the end
+			err = rmErr
+		}
+	}
+
+	return err
+}
+
+// cleanSnapDownloads removes downloads that are no longer needed for the given snap.
+//
+// It is the caller's responsibility to lock state before calling this function.
+var cleanSnapDownloads = func(st *state.State, snapName string) error {
+	keep, err := downloadsToKeep(st)
+	if err != nil {
+		return err
+	}
+
+	regex := regexp.MustCompile(fmt.Sprintf("^%s_x?[0-9]+\\.snap$", snapName))
+
+	matches, err := filepath.Glob(filepath.Join(dirs.SnapBlobDir, fmt.Sprintf("%s_*.snap", snapName)))
```

Comment:
This can only error if the pattern is malformed so I think it makes sense to immediately return the error without going through the matches (even if it's the same in practice, it seems clearer to me)

Reply:
My bad, I actually forgot to add error handling here. Great catch, thank you!